wfcGen.py

```diff
@@ -0,0 +1,287 @@
+            
+            if flag != -1:
+                if WM:
+                    cells[flag]["nbg"]["0"].append(cropped[ins][y][x-1-(res-1)])
+                    cells[flag]["nbg"]["1"].append(cropped[ins][y-1-(res-1)][x])
+                    cells[flag]["nbg"]["2"].append(cropped[ins][y][(x+1+(res-1))%lcx])
+                    cells[flag]["nbg"]["3"].append(cropped[ins][(y+1+(res-1))%lcy][x])
+                else:
+                    cells[flag]["nbg"]["0"].append(cropped[ins][y][x-1])
+                    cells[flag]["nbg"]["1"].append(cropped[ins][y-1][x])
+                    cells[flag]["nbg"]["2"].append(cropped[ins][y][(x+1)%lcx])
+                    cells[flag]["nbg"]["3"].append(cropped[ins][(y+1)%lcy][x])
+            else:
+                if WM:
+                    cells[str(len(cells.keys()))] = {}
+                    cells[str(len(cells.keys())-1)]["nbg"] = {}
+                    cells[str(len(cells.keys())-1)]["nbg"]["0"] = [cropped[ins][y][x-1-(res-1)]]
+                    cells[str(len(cells.keys())-1)]["nbg"]["1"] = [cropped[ins][y-1-(res-1)][x]]
+                    cells[str(len(cells.keys())-1)]["nbg"]["2"] = [cropped[ins][y][(x+1+(res-1))%lcx]]
+                    cells[str(len(cells.keys())-1)]["nbg"]["3"] = [cropped[ins][(y+1+(res-1))%lcy][x]]
+                    cells[str(len(cells.keys())-1)]["img"] = cropped[ins][y][x]
+                else:
+                    cells[str(len(cells.keys()))] = {}
+                    cells[str(len(cells.keys())-1)]["nbg"] = {}
+                    cells[str(len(cells.keys())-1)]["nbg"]["0"] = [cropped[ins][y][x-1]]
+                    cells[str(len(cells.keys())-1)]["nbg"]["1"] = [cropped[ins][y-1][x]]
+                    cells[str(len(cells.keys())-1)]["nbg"]["2"] = [cropped[ins][y][(x+1)%lcx]]
+                    cells[str(len(cells.keys())-1)]["nbg"]["3"] = [cropped[ins][(y+1)%lcy][x]]
+                    cells[str(len(cells.keys())-1)]["img"] = cropped[ins][y][x]
+
+    for y in range(lcy):
+        for x in range(lcx):
+            for key in cells.keys():
+                if cells[key]["img"] == cropped[ins][y][x]:
+                    Gkey = key
+            for key in cells.keys():
+                for i in range(len(cells[key]["nbg"]["0"])):
+                    if cells[key]["nbg"]["0"][i] == cropped[ins][y][x]:
+                        cells[key]["nbg"]["0"][i] = Gkey
+                for i in range(len(cells[key]["nbg"]["1"])):
+                    if cells[key]["nbg"]["1"][i] == cropped[ins][y][x]:
+                        cells[key]["nbg"]["1"][i] = Gkey
+                for i in range(len(cells[key]["nbg"]["2"])):
+                    if cells[key]["nbg"]["2"][i] == cropped[ins][y][x]:
+                        cells[key]["nbg"]["2"][i] = Gkey
+                for i in range(len(cells[key]["nbg"]["3"])):
+                    if cells[key]["nbg"]["3"][i] == cropped[ins][y][x]:
+                        cells[key]["nbg"]["3"][i] = Gkey

```

Comment:
Why is this done separately from the previous block of for loops? Is there a technical reason? Is it for the sake of code readability?

Reply:
yes it is for technical reason

Reply:
and i won't fix it

Reply:
What kind of reason? Why will you not fix it?

Reply:
Alright. In that case, I wonder, if you can rewrite this code such that instead of having four different for loops (for each of the `cells[key]["nbg"]["0"]`, `["1"]`, `["2"]`, `["3"]`) you have a general for loop through all keys/values of `cells[key]["nbg"]`.

By the way, I can see you are doing something like
```python
for i in range(len(some_list)):
    do_something_with(some_list[i])
```
So you are iterating through indices of `some_list` but never using the index directly, only the corresponding element. In such cases it is better to write
```python
for elem in some_list:
    do_something_with(elem)
```